File: LEETCODE/LEETCODE-CPP/49.cpp
//49. Group Anagrams
//https://leetcode.com/problems/group-anagrams/description/


//Method-1 using hashmap and sorting
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int n=strs.size();
        unordered_map<string,vector<string>>mp;
        vector<vector<string>>ans;
        if(n<0){
            return ans;
        }

        for(int i=0;i<n;i++){
            string temp=strs[i];

            sort(strs[i].begin(),strs[i].end());
            mp[strs[i]].push_back(temp);
        }

        for(auto i=mp.begin();i!=mp.end();i++){
            ans.push_back(i->second);
        }

        return ans;
    }
};


//Method-1 using hashmap and without sorting
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int n=strs.size();
        unordered_map<string,vector<string>>mp;
        vector<vector<string>>ans;
        if(n<0){
            return ans;
        }

        for(auto s:strs){
            vector<int> arr(26,0);
            for(auto x:s){
                arr[x-'a']++;
            }
            
            string temp="";
            for(int i=0;i<26;i++){
                temp+=to_string(arr[i]);
                temp=temp+'0';
            }

            mp[temp].push_back(s);
        }

        for(auto i=mp.begin();i!=mp.end();i++){
            ans.push_back(i->second);
        }

        return ans;
    }
};

File: LEETCODE/LEETCODE-PYTHON/49.py
# 49. Group Anagrams
# https://leetcode.com/problems/group-anagrams/description/

# Method-1 categorize by sorting string O(NKlogK)
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res=collections.defaultdict(list)

        for s in strs:
            res[tuple(sorted(s))].append(s)

        return list(res.values())

# Method-2 count character frequence O(NK)
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res=collections.defaultdict(list)

        for s in strs:
            temp=[0]*26

            for i in s:
                temp[ord(i)-ord('a')]+=1

            res[tuple(temp)].append(s)

        return list(res.values())